### PR TITLE
Gift Code Memos

### PR DESCRIPTION
**Pull request by iamalwaysuncomfortable**
### Motivation

As a part of [MCIP 32](https://github.com/mobilecoinfoundation/mcips/blob/main/text/0032-fog-compatible-gift-codes.md), 3 new memo types were specified:

| Memo type bytes | Name                                              |
| -----------     | -----------                                       |
| 0x0201          | Gift code funding memo                            |
| 0x0202          | Gift code cancellation memo                       |
| 0x0002          | Gift code sender memo                       |

This pull request implements the memo types for the MCIP.  Specific questions exist about whether the gift code funding memo should involve constant time operations when validating the TxOut hash.

### In this PR
Implementation of the 3 memo types specified in MCIP 32

Addresses work specified in #1735 

### Future Work
* Implement gift code memo builder object
* Implement git code protobuf object
* Implement gift code transactions within the transaction builder
* Build codepaths for sending gift codes within libmobilecoin and mobilecoind

### CODE REVIEW

File: transaction/std/src/memo/gift_code_funding.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+impl RegisteredMemoType for GiftCodeFundingMemo {
+    const MEMO_TYPE_BYTES: [u8; 2] = [0x02, 0x01];
+}
+
+impl GiftCodeFundingMemo {
+    /// Create a new gift funding code memo
+    pub fn new(tx_out_public_key: RistrettoPublic, note: &str) -> Result<Self, MemoError> {
+        // Check if note is of valid length and initialize memo data
+        if note.len() > Self::NOTE_DATA_LEN {
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_funding.rs:36]: This doesn't leave room for the null terminator. We could make it optional if the string is exactly 60 bytes long but that is a bit non conventional.

File: transaction/std/src/memo/gift_code_funding.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    /// Length of the utf-8 note
+    pub const NOTE_DATA_LEN: usize = 60;
+
+    /// Get the memo data
+    pub fn memo_data(&self) -> &[u8; Self::MEMO_DATA_LEN] {
+        &self.memo_data
+    }
+
+    /// Check if a given public key matches
+    /// TODO: Should this be constant time?
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_funding.rs:67]: I dont see why, this is only done on clients and we have not been doing any constant time work there as far as I know.

File: transaction/std/src/memo/gift_code_funding.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    pub fn public_key_matches(&self, tx_out_public_key: RistrettoPublic) -> bool {
+        tx_out_public_key_short_hash(tx_out_public_key) == self.memo_data[0..4]
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_funding.rs:69]: This should use HASH_DATA_LEN

File: transaction/std/src/memo/gift_code_funding.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /// Get funding note from memo
+    pub fn funding_note(&self) -> &str {
+        let note_data = &self.memo_data[Self::HASH_DATA_LEN..];
+        str::from_utf8(note_data)
+            .unwrap()
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_funding.rs:76]: This should not unwrap, it will panic if the memo has an invalid utf8 sequence. Instead, this method should return a Result

File: transaction/std/src/memo/gift_code_funding.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            .trim_matches(char::from(0))
+    }
+}
+
+// Compute first four bytes of TxOut hash
+fn tx_out_public_key_short_hash(
+    tx_out_public_key: RistrettoPublic,
+) -> [u8; GiftCodeFundingMemo::HASH_DATA_LEN] {
+    let mut hasher = Blake2b512::new();
+    hasher.update("mc-memo-mac");
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_funding.rs:86]: Nit: I think this string should be more specific to what is being hashed here (e.g. `mc-gift-funding-tx-pub-key`). It definitely shouldn't match another hash in the codebase.

File: transaction/std/src/memo/gift_code_funding.rs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    hasher.update(tx_out_public_key.as_ref().compress().as_bytes());
+    hasher.finalize().as_slice()[0..GiftCodeFundingMemo::HASH_DATA_LEN]
+        .try_into()
+        .unwrap()
+}
+
+impl From<&[u8; 64]> for GiftCodeFundingMemo {
+    fn from(src: &[u8; Self::MEMO_DATA_LEN]) -> Self {
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_funding.rs:94]: If we are enforcing that the memo is null-terminated, this should potentially get checked here.

File: transaction/std/src/memo/gift_code_sender.rs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+//! Object for 0x0002 Gift Code Sender memo type
+//!
+//! This was proposed for standardization in mobilecoinfoundation/mcips/pull/32
+
+use crate::{impl_memo_type_conversions, RegisteredMemoType};
+use mc_transaction_core::MemoError;
+use std::{convert::TryFrom, str};
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct GiftCodeSenderMemo {
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_sender.rs:12]: Can you please include a comment explaining what this memo is like you did for the other ones?

**Review by eranrund — CHANGES_REQUESTED** [review]

File: transaction/std/src/memo/gift_code_cancellation.rs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        &self.memo_data
+    }
+
+    /// Get global TxOut index of the cancelled gift code
+    pub fn cancelled_gift_code_index(&self) -> u64 {
+        u64::from_le_bytes(self.memo_data[0..8].try_into().unwrap())
+    }
+}
+
+impl From<&[u8; 64]> for GiftCodeCancellationMemo {
```
Comment (remoun) [inline transaction/std/src/memo/gift_code_cancellation.rs:45]: Let's use MEMO_DATA_LEN in the type, too.
```suggestion
impl From<&[u8; GiftCodeCancellationMemo::MEMO_DATA_LEN]> for GiftCodeCancellationMemo {
```

File: transaction/std/src/memo/gift_code_funding.rs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    pub const NOTE_DATA_LEN: usize = 60;
+
+    /// Get the memo data
+    pub fn memo_data(&self) -> &[u8; Self::MEMO_DATA_LEN] {
+        &self.memo_data
+    }
+
+    /// Check if a given public key matches
+    /// TODO: Should this be constant time?
+    pub fn public_key_matches(&self, tx_out_public_key: RistrettoPublic) -> bool {
```
Comment (remoun) [inline transaction/std/src/memo/gift_code_funding.rs:68]: Nit: pass by ref
```suggestion
    pub fn public_key_matches(&self, tx_out_public_key: &RistrettoPublic) -> bool {
```

File: transaction/std/src/memo/gift_code_funding.rs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+) -> [u8; GiftCodeFundingMemo::HASH_DATA_LEN] {
+    let mut hasher = Blake2b512::new();
+    hasher.update("mc-memo-mac");
+    hasher.update(tx_out_public_key.as_ref().compress().as_bytes());
+    hasher.finalize().as_slice()[0..GiftCodeFundingMemo::HASH_DATA_LEN]
+        .try_into()
+        .unwrap()
+}
+
+impl From<&[u8; 64]> for GiftCodeFundingMemo {
```
Comment (remoun) [inline transaction/std/src/memo/gift_code_funding.rs:93]:
```suggestion
impl From<&[u8; GiftCodeFundingMemo::MEMO_DATA_LEN]> for GiftCodeFundingMemo {
```

File: transaction/std/src/memo/gift_code_cancellation.rs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    fn from(src: &[u8; Self::MEMO_DATA_LEN]) -> Self {
+        let mut memo_data = [0u8; Self::MEMO_DATA_LEN];
+        memo_data.copy_from_slice(src);
+        Self { memo_data }
+    }
+}
+
+impl From<GiftCodeCancellationMemo> for [u8; GiftCodeCancellationMemo::MEMO_DATA_LEN] {
```
Comment (remoun) [inline transaction/std/src/memo/gift_code_cancellation.rs:53]: How about we make these `From<Gift*Memo>` passed by reference? It seems inconsistent to make this From impl consume its param, while the other direction (from `&[u8; 64]`) passed by reference

File: transaction/std/src/memo/gift_code_funding.rs (hunk 1 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:36]: Yeah I think we can probably make the null terminator mandatory for consistency

File: transaction/std/src/memo/gift_code_funding.rs (hunk 3 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:69]: Yes it should

File: transaction/std/src/memo/gift_code_funding.rs (hunk 2 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:67]: I was thinking maybe because it's potentially sensitive information (the attacker finding out specific txout that the gift code is if they've managed to intercept the message with the spend). But probably such an edge case, and only happens once per gift code tx, that we don't have to worry about it.

File: transaction/std/src/memo/gift_code_funding.rs (hunk 4 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:76]: Also YES

File: transaction/std/src/memo/gift_code_funding.rs (hunk 5 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:86]: So.. here's what I'm thinking about this, and give me your opinion. The Domain separators in the core crate are not public and can't be accessed here - does it make sense to create a separate set of domain separators for the transaction/std crate?

File: transaction/std/src/memo/gift_code_funding.rs (hunk 6 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:94]: Agreed - more tests needed

File: transaction/std/src/memo/gift_code_funding.rs [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+impl From<GiftCodeFundingMemo> for [u8; GiftCodeFundingMemo::MEMO_DATA_LEN] {
+    fn from(src: GiftCodeFundingMemo) -> [u8; GiftCodeFundingMemo::MEMO_DATA_LEN] {
+        src.memo_data
+    }
+}
+
+impl_memo_type_conversions! { GiftCodeFundingMemo }
+
+#[cfg(test)]
+mod tests {
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:110]: Noted

File: transaction/std/src/memo/gift_code_sender.rs [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        GiftCodeSenderMemo { memo_data }
+    }
+
+    /// Get the memo data
+    pub fn memo_data(&self) -> &[u8; Self::MEMO_DATA_LEN] {
+        &self.memo_data
+    }
+
+    /// Get the sender note
+    pub fn sender_note(&self) -> &str {
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_sender.rs:36]: Noted

File: transaction/std/src/memo/gift_code_sender.rs [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        let mut memo_data = [0u8; Self::MEMO_DATA_LEN];
+        memo_data[0..src.len()].copy_from_slice(src.as_bytes());
+        Ok(Self { memo_data })
+    }
+}
+
+impl_memo_type_conversions! { GiftCodeSenderMemo }
+
+#[cfg(test)]
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_sender.rs:72]: Will do

File: transaction/std/src/memo/gift_code_funding.rs (hunk 9 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:68]: Agreed

File: transaction/std/src/memo/gift_code_funding.rs (hunk 5 above)
Comment (eranrund) [reply transaction/std/src/memo/gift_code_funding.rs:86]: No reason not to create a domain separators file. But I also don't see much value in it because they only get used in a single place, it's simple enough to ensure this is unique and specific enough when it gets added.

File: transaction/std/src/memo/gift_code_cancellation.rs (hunk 8 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_cancellation.rs:45]: The difference was that I could see potentially MEMO_DATA_LEN gets shorter than 64 bytes and thus breaks the trait. I think it should be okay though unless someone wants this MEMO_DATA_LEN to be shorther.

File: transaction/std/src/memo/gift_code_sender.rs [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+    /// Get the sender note
+    pub fn sender_note(&self) -> &str {
+        str::from_utf8(&self.memo_data)
+            .unwrap()
+            .trim_matches(char::from(0))
+    }
+}
+
+impl From<&[u8; 64]> for GiftCodeSenderMemo {
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_sender.rs:43]: Yeah, if no one has nits on the MEMO_DATA_LEN being anything but 64, this will work

File: transaction/std/src/memo/gift_code_cancellation.rs (hunk 11 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_cancellation.rs:53]: Yeah actually that's what the trait specifies

Comment (remoun) [reply transaction/std/src/memo/gift_code_cancellation.rs:53]: I mean `From<&GiftCodeCancellationMemo>` (with &)

Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_cancellation.rs:53]: Because that would break the Into trait for RegisteredMemoType trait as designed currently by enforcing a lifetime on the input (I think). The src can be a &GiftCodeCancellationMemo even when the trait is the owned type, which I should have done!

i.e. it should be:
```rust
impl From<GiftCodeCancellationMemo> for [u8; GiftCodeCancellationMemo::MEMO_DATA_LEN] {
    fn from(src: GiftCodeCancellationMemo) -> [u8; GiftCodeCancellationMemo::MEMO_DATA_LEN] {
        src.memo_data
    }
}
```

Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_cancellation.rs:53]: Because it breaks the Into trait. It would be better to pass from reference, but that breaks the trait somehow. To be quite honest, I'm not sure why.

Comment (eranrund) [reply transaction/std/src/memo/gift_code_cancellation.rs:53]: What do you mean by it breaks the trait?

File: transaction/std/src/memo/gift_code_funding.rs (hunk 6 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:94]: To enforce it - I think the only option here is potentially truncate any possible last non-null character in order to keep the trait bound consistent. Null termination was in the mcip but I'm not necessarily sure we need to null terminate the strings.

File: transaction/std/src/memo/gift_code_cancellation.rs (hunk 11 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_cancellation.rs:53]: It complains about Into not being satisfied for the RegisteredMemoType trait:
`the trait bound `GiftCodeCancellationMemo: Into<[u8; 64]>` is not satisfied [E0277] the trait `Into<[u8; 64]>` is not implemented for `GiftCodeCancellationMemo``

Comment (eranrund) [reply transaction/std/src/memo/gift_code_cancellation.rs:53]: Yup, looks like `RegisteredMemoType` requires it: `pub trait RegisteredMemoType:
    Sized + Clone + Debug + Into<[u8; 64]> + for<'a> From<&'a [u8; 64]>`

File: transaction/std/src/memo/gift_code_funding.rs (hunk 1 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:36]: Actually, after thinking some, I think it should be a string that's "null terminated or all 64 bytes".

Namely because, in my opinion:
1. Clients then get to use all bytes
2. In both utf-8 cases, the byte length reserved for utf-8 is fixed so "stop at the first null byte or use all the string" is fairly straightforward
3. Most of the langs that'd consume this (Java/Swift/Dart/JS/Rust) also don't use null terminated strings

File: transaction/std/src/memo/gift_code_funding.rs (hunk 6 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:94]: Also in general, we should maybe consider making the RegisteredMemoType `TryFrom` instead of `From`. It seems like it would allow us to handle bad from byte conversions with an Error Monad instead of a panic. Thoughts?

Comment (eranrund) [reply transaction/std/src/memo/gift_code_funding.rs:90]: We could also keep it this way, since this is just storing the 64 bytes, and have the error checking when specific fields are accessed. @garbageslam I wonder if you have any thoughts on this.

File: transaction/std/src/memo/gift_code_funding.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:109]: Done! Went forward with the assumption that the last byte can be non-null if the length is 60 for funding/64 for sender. 

If we all agree on this, I'll PR a change into MCIP 32 stating this.

File: transaction/std/src/memo/gift_code_funding.rs [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+        let note = "Cash money MeowbleCoin for Kitty";
+        let txout_public_key = RistrettoPublic::from_random(&mut rng);
+        let memo = GiftCodeFundingMemo::new(txout_public_key, note).unwrap();
+
+        // Check that the note is extracted properly
+        assert_eq!(memo.funding_note(), note);
+
+        // Check that the public key can be verified
+        assert!(memo.public_key_matches(txout_public_key));
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs]: Done!

File: transaction/std/src/memo/gift_code_sender.rs [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+impl RegisteredMemoType for GiftCodeSenderMemo {
+    const MEMO_TYPE_BYTES: [u8; 2] = [0x00, 0x02];
+}
+
+impl GiftCodeSenderMemo {
+    /// The length of the custom memo data.
+    pub const MEMO_DATA_LEN: usize = 64;
+
+    /// Create a new gift code memo
+    pub fn new(note_data: &'static str) -> Result<Self, MemoError> {
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_sender.rs:34]: What do I do if I want to attach a memo that is not a static string, e.g. something that came from a `String`? This should be `&str`

Comment (eranrund) [reply transaction/std/src/memo/gift_code_sender.rs:34]: I'm also not sure we need both `new` and a TryFrom

File: transaction/std/src/memo/gift_code_funding.rs [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+        let key = RistrettoPublic::from_random(&mut rng);
+
+        // Create memos from notes
+        let memo_len_minus_one = GiftCodeFundingMemo::new(&key, note_len_minus_one).unwrap();
+        let memo_len_exact = GiftCodeFundingMemo::new(&key, note_len_exact).unwrap();
+        let memo_len_plus_one = GiftCodeFundingMemo::new(&key, note_len_plus_one);
+
+        // Check note lengths match or error on creation if note is too large
+        let _memo_err: Result<GiftCodeFundingMemo, MemoError> =
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_funding.rs:223]: Nit: underscore prefix is for unused variables

File: transaction/std/src/memo/gift_code_sender.rs [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+        let note_len_plus_one =
+            str::from_utf8(&[b'6'; GiftCodeSenderMemo::MEMO_DATA_LEN + 1]).unwrap();
+
+        // Create memos from notes
+        let memo_len_minus_one = GiftCodeSenderMemo::new(note_len_minus_one).unwrap();
+        let memo_len_exact = GiftCodeSenderMemo::new(note_len_exact).unwrap();
+        let memo_len_plus_one = GiftCodeSenderMemo::new(note_len_plus_one);
+
+        // Check note lengths match or error on creation if too large
+        let _memo_err: Result<GiftCodeSenderMemo, MemoError> =
```
Comment (eranrund) [inline transaction/std/src/memo/gift_code_sender.rs:150]: Nit: underscore prefix is for unused variables

**Review by eranrund — CHANGES_REQUESTED** [review]
A few minor things left but other than that LGTM.
Nice work with all the tests, thank you!

File: fog/sample-paykit/src/cached_tx_data/memo_handler.rs [hunk 20]
```diff
@@ -108,6 +108,18 @@ impl MemoHandler {
                     Err(MemoHandlerError::FailedSubaddressValidation)
                 }
             }
+            MemoType::GiftCodeCancellation(_) => {
+                // TODO: Add Gift Code Memo Validation
```
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply fog/sample-paykit/src/cached_tx_data/memo_handler.rs:112]: Good point, tracked: https://github.com/mobilecoinfoundation/mobilecoin/issues/1959

File: transaction/std/src/memo/gift_code_funding.rs (hunk 18 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:223]: Yes, while this is true, I used it because
1. The compiler didn't consider the usage within the corresponding macro below an actual usage
2. The memo didn't allow addition within a nested macro, and I wanted to avoid putting 61/65 directly. I could've potentially used assert_matches! but that's nightly rust and James has expressed the desire to move away from nightly rust - so this was the compromise :)

File: transaction/std/src/memo/gift_code_sender.rs (hunk 19 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_sender.rs:150]: ^^ Same as above, maybe there's a different way?

File: transaction/std/src/memo/gift_code_funding.rs (hunk 18 above)
Comment (eranrund) [reply transaction/std/src/memo/gift_code_funding.rs:223]: That is quite funky. I think a potential solution is something like this:
`assert!(matches!(memo_len_plus_one, Err(MemoError::BadLength(len)) if len == GiftCodeFundingMemo::NOTE_DATA_LEN + 1)`
Would you mind trying that please?

File: transaction/core/src/memo.rs [hunk 21]
```diff
@@ -221,6 +224,15 @@ derive_prost_message_from_repr_bytes!(MemoPayload);
 pub enum MemoError {
     /// Wrong length for memo payload: {0}
     BadLength(usize),
+
+    /// Utf-8 did note properly decode
```
[the remark this answers is not in the document]
Comment (eranrund) [reply transaction/core/src/memo.rs]: This is still pending

File: transaction/std/src/memo/gift_code_funding.rs (hunk 18 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs]: That failed, but found an alternate solution.

File: transaction/std/src/memo/gift_code_sender.rs (hunk 19 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_sender.rs]: Found an alternate solution

File: transaction/core/src/memo.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply transaction/core/src/memo.rs]: Apologies, passed now.

**Review by eranrund — APPROVED** [review]
LGTM. Would like to hear what other reviewers think about making the null terminator optional when the entire length is used.

File: fog/sample-paykit/src/cached_tx_data/memo_handler.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (garbageslam) [reply fog/sample-paykit/src/cached_tx_data/memo_handler.rs:112]: @iamalwaysuncomfortable I think the MCIP doesn't specify any way to validate the memos, so this is the final state right?

File: transaction/std/src/memo/gift_code_cancellation.rs [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+impl GiftCodeCancellationMemo {
+    /// The length of the custom memo data
+    pub const MEMO_DATA_LEN: usize = 64;
+
+    /// Create a new gift code memo
+    pub fn new(global_index: u64) -> Self {
+        GiftCodeCancellationMemo::from(global_index)
+    }
+
+    /// Get the memo data
```
Comment (garbageslam) [inline transaction/std/src/memo/gift_code_cancellation.rs:34]: So this is not the most useful API to expose here.

The `mc_transaction_core::MemoPayload` type gives the user access to the types. The `GiftCodeCancellationMemo` struct is supposed to be like a higher level type that abstracts away the details of the binary format.

So for instance, I would want to see like:

```
pub fn global_index(&self) -> u64;
pub fn note(&self) -> &str;
```

rather than just handing the user the entire Memo data, because that's not very useful to the SDKs and users, they would all have to refer to the MCIP to see how to interpret that data and we would get a ton of code duplication

Comment (garbageslam) [reply transaction/std/src/memo/gift_code_cancellation.rs:34]: I see you have the `cancelled_gift_code_index` one, but I think we should get rid of the `memo_data` and make it just return the note, also checking that it's valid utf8

File: transaction/std/src/memo/gift_code_funding.rs [hunk 23]
```diff
@@ -0,0 +1,310 @@
...
+            return Ok(note.0);
+        }
+        Ok(note)
+    }
+}
+
+// Compute first four bytes of TxOut hash
+fn tx_out_public_key_short_hash(
+    tx_out_public_key: &RistrettoPublic,
+) -> [u8; GiftCodeFundingMemo::HASH_DATA_LEN] {
```
Comment (garbageslam) [inline transaction/std/src/memo/gift_code_funding.rs:83]: this looks good, thank you

File: transaction/std/src/memo/gift_code_funding.rs [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+    }
+
+    /// Check if a given public key matches
+    pub fn public_key_matches(&self, tx_out_public_key: &RistrettoPublic) -> bool {
+        tx_out_public_key_short_hash(tx_out_public_key) == self.memo_data[0..Self::HASH_DATA_LEN]
+    }
+
+    /// Get funding note from memo
+    pub fn funding_note(&self) -> Result<&str, MemoError> {
+        let note = str::from_utf8(&self.memo_data[Self::HASH_DATA_LEN..])?;
```
Comment (garbageslam) [inline transaction/std/src/memo/gift_code_funding.rs:72]: I think this is not the best way to implement this, because it's going to stick a bunch of null characters in the resulting string:

https://users.rust-lang.org/t/should-string-from-utf8-remove-trailing-null/52680

it would be better to do something where we find the index of the first null, if any, then change this to be like

```
let note = str::from_utf8(&self.memo_data[Self::HASH_DATA_LEN..first_null_index])?;
```

Comment (garbageslam) [reply transaction/std/src/memo/gift_code_funding.rs:72]: although maybe i'm confused, because your unit test `test_gift_code_funding_memo_data_outputs_match()` seems to be passing...

Comment (garbageslam) [reply transaction/std/src/memo/gift_code_funding.rs:72]: oh i see, you are doing the split_once thing after converting to utf-8

i think it might be marginally better to find the null before converting to utf-8, because it's conceivable that the stuff after the first null is not valid utf-8. but it's probably fine as is, i guess we'll always zero the unused bytes.

File: transaction/std/src/memo/gift_code_funding.rs [hunk 25]
```diff
@@ -0,0 +1,310 @@
...
+
+        // Check that the note is extracted properly
+        assert_eq!(memo.funding_note().unwrap(), note);
+
+        // Check that the public key can be verified
+        assert!(memo.public_key_matches(&key));
+    }
+
+    #[test]
+    fn test_gift_code_funding_memo_with_blank_note_is_ok() {
```
Comment (garbageslam) [inline transaction/std/src/memo/gift_code_funding.rs:130]: Can we also have a test of what happens when the memo is the maximum length 56 bytes?

Comment (garbageslam) [reply transaction/std/src/memo/gift_code_funding.rs:130]: nm you already did this :)

**Review by garbageslam — APPROVED** [review]
LGTM except that I think the cancellation memo should have an accessor that gives `&str` like the others. thanks for doing this and writing good tests

File: fog/sample-paykit/src/cached_tx_data/memo_handler.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (iamalwaysuncomfortable) [reply fog/sample-paykit/src/cached_tx_data/memo_handler.rs:112]: Maybe? I think if the FOG paykit supports testing gift code transactions, we could validate that they arrive properly at least

File: transaction/std/src/memo/gift_code_funding.rs (hunk 24 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_funding.rs:72]: Actually yeah, that IS the better way, I'm going to refactor to that

File: transaction/std/src/memo/gift_code_cancellation.rs (hunk 22 above)
Comment (iamalwaysuncomfortable) [reply transaction/std/src/memo/gift_code_cancellation.rs:34]: The gift code cancellation txout actually doesn't have a utf-8 byte specification, in the MCIP - it's "unused". But I could remove the method which returns the bytes

**Comment by garbageslam** [comment]
> LGTM. Would like to hear what other reviewers think about making the null terminator optional when the entire length is used.

This is how it currently works right? It seems good to me to just give the user as much space as possible. I don't think the "conditional" null terminator is bad, it's basically the same as `strnlen` and I don't think this is a bad practice, the null-terminator search is bounded here.

File: transaction/std/src/memo/gift_code_sender.rs [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+            memo_len_plus_one,
+            Err(MemoError::BadLength(LEN_PLUS_ONE))
+        ));
+    }
+
+    #[test]
+    fn test_gift_code_sender_memo_with_corrupted_bytes_fails() {
+        // Create note bytes and corrupt them
+        let mut note_bytes = [b'6'; GiftCodeSenderMemo::MEMO_DATA_LEN - 1];
+        let note = str::from_utf8(&[b'6'; GiftCodeSenderMemo::MEMO_DATA_LEN - 1]).unwrap();
```
Comment (remoun) [inline transaction/std/src/memo/gift_code_sender.rs:163]: This should be `str::from_utf(&note_bytes)`
